server/src/main/java/org/opensearch/cluster/service/TaskBatcher.java

```diff
@@ -139,6 +155,13 @@ private void onTimeoutInternal(List<? extends BatchedTask> tasks, TimeValue time
                 }
                 return existingTasks;
             });
+            taskIdentityPerBatchingKey.computeIfPresent(batchingKey, (tasksKey, existingIdentities) -> {
```

Comment:
We are maintaining specific order in updating both maps, as identity map will be updated first and removed last. 

For task submission, we will first insert the task identity in map and then insert task in map. Here out of sync will not cause issue as identity will be inserted first, so let's say even if tasks are not inserted in map and duplicate tasks are submitted again, it will be rejected from identity map itself.

Similarly, for remove case[execute/before timeout] we will need to remove tasks from map first and then remove the task identity from map. Here out of sync can fail duplicate tasks which are already removed from tasks map due to execute/before timeout scenario. Since task is already going to execute/timeout, it should be safe to reject duplicate task.

With specific order in which maps are updated, we might not require synchronize block, that's why I haven't put the synchronize block.

On another hands, now de-dup logic has improved so operation in synchronize block will not take much time but we will need to put block at map level, which was earlier case before we move to concurrent hash map. This might impact performance a bit.

Let me know your thoughts or incase I am missing something.

Reply:
> On another hands, now de-dup logic has improved so operation in synchronize block will not take much time but we will need to put block at map level, which was earlier case before we move to concurrent hash map. This might impact performance a bit.

which block is this. we dont need lock the whole map at any point.

Reply:
I see your point on maintaining the order in which maps are updated. I think it is reasonable. Please add that in the comment and mention not the change the map update order otherwise it will result in duplicates.

Reply:
Added comment in code for this.